### ISSUE TITLE
PASM debug not correctly ignored when containing immediates

### ISSUE THREAD

**Issue by Wuerfel21** [totalspectrum/spin2cpp#182]
While looking into the PASM debug thing, I noticed that the parser fails if the debug exprlist contains an immediate (`#.loop`)
```spin
CON
_CLKFREQ = 10_000_000

DAT

.loop
mov pa,#0
debug(udec(pa),udec(#.loop),"speck")
add pa,#1
jmp #.loop
```

If I get it actually working, I'll have fixed the problem along the way and you don't need to do anything.

**Comment by Wuerfel21** [comment]
Kinda unrelatedly, I have to assign each DEBUG a unique BRK code. I tried doing it in `AssignAddresses`, but that seems to get called twice with a different copy of the AST, so that doesn't work. Where should that happen then? I guess I could put it in `PrintDataBlock` ??

**Comment by totalspectrum** [comment]
I'd probably emit a new AST_PASM_DEBUG node, and then assign a unique code to it in PrintDataBlock. AssignAddresses is a bit of a hack. Actually the whole assembly / DAT section is messy -- it's some of the oldest part of the code, and the parser doesn't output a clean list for it the way we do in the rest of the code. I've been meaning to refactor it for a long time, but the last time I tried I broke things spectacularly and gave up...

**Comment by Wuerfel21** [comment]
> I'd probably emit a new AST_PASM_DEBUG node, and then assign a unique code to it in PrintDataBlock
Yeah, that's what I'm doing / going to do.

**Comment by Wuerfel21** [comment]
Also, how would I best go about prepending the debugger and its data table in front of the program?

**Comment by totalspectrum** [comment]
Does the program itself have to be modified (e.g. compiled at a different address) or is it just a matter of concatenating the two binaries? If it's just sticking the binaries together it would probably go where DoPropellerChecksum goes now (perhaps we'd replace it with DoPropellerPostprocess that does both concat and checksum, where appropriate).

If instead we need to compile the program with different addresses, that's perhaps a bit trickier. If it's a fixed offset then we could do the same as above, with the gl_dat_offset set appropriately. If a variable offset, then probably it needs to go in the assembly generation in backends/asm/assemble_ir.c.

**Comment by Wuerfel21** [comment]
Yeah, it just needs to be concatenated in front of the binary. I already thought of doing it there, problem is that `DoPropellerChecksum` patches the on-disk file in `r+b` mode, which isn't really suited to prepending data.

**Comment by Wuerfel21** [comment]
I guess I'll try to refactor it to run in-memory?

**Comment by Wuerfel21** [comment]
And of course I did not do that and instead resorted to this terribleness
```c
    uint32_t extralen = 0;
    if (gl_brkdebug) {
        char buffer[len]; // allocate VLA
        fseek(f,0,SEEK_SET);
        fread(buffer,len,1,f);
        Flexbuf debugger = CompileBrkDebugger(len);
        extralen += flexbuf_curlen(&debugger);
        fseek(f,0,SEEK_SET);
        fwrite(flexbuf_peek(&debugger),flexbuf_curlen(&debugger),1,f);
        fwrite(buffer,len,1,f);
        flexbuf_delete(&debugger);
        fflush(f);
    }
```

**Comment by Wuerfel21** [comment]
The actual issue here is fixed by the parser changes in #184
